### ISSUE TITLE
Update builtins for no-extend-native and no-native-reassing

### ISSUE THREAD

**Issue by jrvidal** [eslint/eslint#2972]
There are a few builtins missing in these rules, in particular many of the new ES6 globals.

I already opened a PR in #2971.

**Comment by ilyavolodin** [comment]
I'm a bit concern about your commit because you added a bunch new keywords (for example Reflect) that are not really keywords unless ES6 is enabled. So for somebody who doesn't write ES6 those would be restricted, where in reality - they should not be (one can argue that it's a good idea not to use those in ES5 code anyways, but that's a different discussion).

**Comment by jrvidal** [comment]
Well, the best we could do now is to inspect the global scope to further filter the builtins list. However that wouldn't immediately solve the problem, since the `global` package does not include sufficiently fine-grained info about each global.

And in fact, it's already affecting user code:

```js
var p = new Promise();
```

This code does not trigger `no-undef` regardless of the environment. I'm not sure, but I think that currently there isn't a way for users to disable specific globals. I don't see how we could solve this.

**Comment by ilyavolodin** [comment]
Agree. I would like to hear what others have to say on the subject. I don't think we need to care all that much about ES3, but there are still a lot of ES5 users out there (myself included) and giving them false positives might not be the best idea.
CC @eslint/eslint-team

**Comment by nzakas** [comment]
We already are kind of in that boat with globals already. If you look at the built in list from the `globals` module, it includes ES6 objects.

Related: why wouldn't we just use `globals` for this?

**Comment by ilyavolodin** [comment]
Good point. @jrvidal any issues with just using `globals` for this?

**Comment by jrvidal** [comment]
Fine by me. However we still need a separate list for "constructors" (for `no-extend-native`), if `globals` does not provide that info.

**Comment by ilyavolodin** [comment]
Can't we use `globals` `builtin` for those as well? What would be the problem with that?

**Comment by jrvidal** [comment]
I was concerned that we would trigger false positives on patterns such as:

```js
encodeURI.prototype.foo = 1;
```

where `encodeURI` does not have a prototype to be modified, so if it appears in user code they added it manually. But on second thought it's a pretty strange corner case.

I'll update the PR.

**Comment by nzakas** [comment]
Constructors always Begin with a capital letter, so it should be easy to filter the list.

**Comment by jrvidal** [comment]
There are some exceptions to that:
```js
// With prototype
'constructor', 'clearImmediate', 'clearInterval', 'clearTimeout', 'require', 'setImmediate', 'setInterval', 'setTimeout', 'blur', 'close', 'focus', 'postMessage'

// Without prototype
'Infinity', 'JSON', 'Math', 'GLOBAL', 'CSS', 'Intl'
```

But as I mentioned, they might be too contrived to worry about.

Related: How should we detect which globals to report? I was thinking to expose the globals as configured in `lib/eslint/addDeclaredGlobals` in the `context` object. However, I'm not sure if we should report all of them, in particular for `no-extend-native`. The assumption about upper case Constructors may not hold for explicit globals added by the user in the config file, inline or exported.

**Comment by nzakas** [comment]
Explicit globals aren't native, so they would be exempt from this check.

I'd say go with the naive approach to start. Just use the built-in list from the globals module and lets see what feedback we get.
